### PR TITLE
ENS Registration: Upload Avatar/Cover

### PR DESCRIPTION
**Pull request by jxom**
Fixes RNBW-2563

## What changed (plus any additional context for devs)

This PR adds avatar & cover photo support to ENS registration flow.

I have updated `react-native-image-crop-picker` to the latest version as the current version had issues with our existing cool-modal implementation.

I have updated `react-native-ios-context-menu` to the latest version as the current version was giving me some grievances... Seems like the newer version of the lib also supports TS, which caused the `UniqueTokenExpandedStateHeader` component to throw type errors, so I had to fix them.

## PoW (screenshots / screen recordings)

Kapture 2022-02-22 at 12 23 45

### CODE REVIEW

File: patches/react-native-ios-context-menu+1.7.4.patch [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
++import { TouchableOpacity } from 'react-native-gesture-handler'
+ 
+ import { RNIContextMenuButton, RNIContextMenuButtonBaseProps, RNIContextMenuButtonCommands } from '../../native_components/RNIContextMenuButton';
+ 
+@@ -211,7 +212,9 @@ export class ContextMenuButton extends React.PureComponent<ContextMenuButtonProp
+       ): shouldUseActionSheetFallback? (
+         // C - Use 'ActionSheet' Fallback
+         <TouchableOpacity 
+-          onLongPress={this._handleOnLongPress}
++          {...({
```
Comment (jxom) [inline patches/react-native-ios-context-menu+1.7.4.patch:18]: Just copying over the patch from 1.3.0

Comment (estebanmino) [reply patches/react-native-ios-context-menu+1.7.4.patch:18]: I've been trying this in other places we're using context menu and is looking good

File: src/components/ens-registration/Cover/Cover.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  Cover,
+  Text,
+  useForegroundColor,
+} from '@rainbow-me/design-system';
+import { useENSProfileForm, useSelectImageMenu } from '@rainbow-me/hooks';
+import { ImgixImage } from '@rainbow-me/images';
+
+const alpha = '33';
+
+export default function CoverPhoto() {
```
Comment (estebanmino) [inline src/components/ens-registration/Cover/Cover.tsx:17]: maybe is this file maybe not but I couldn't go to the place where I could choose a picture for cover, I was able to choose and image for pfp thi

File: src/components/ens-registration/Avatar/Avatar.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    },
+    uploadToIPFS: true,
+  });
+
+  return (
+    <Box height={{ custom: size }} width={{ custom: size }}>
+      <Cover alignHorizontal="center">
+        <BackgroundProvider color="body">
+          {({ backgroundColor }) => (
+            <Svg height="32" style={{ top: -6 }} viewBox="0 0 96 29" width="96">
```
Comment (estebanmino) [inline src/components/ens-registration/Avatar/Avatar.tsx:63]: could we move this one to `src/components/icons/svg/` just to have all svgs there

File: src/screens/ENSAssignRecordsSheet.js [hunk 4]
```diff
@@ -89,63 +73,10 @@ export default function ENSAssignRecordsSheet() {
...
-                      <AccentColorProvider color={accentColor}>
-                        <Text color="accent" size="18px" weight="heavy">
-                          {` 􀣵 `}
-                        </Text>
-                      </AccentColorProvider>
-                    </Box>
-                  </AccentColorProvider>
-                </ButtonPressAnimation>
-              </Box>
+              <Avatar onChangeAvatarUrl={setAvatarUrl} />
```
Comment (estebanmino) [inline src/screens/ENSAssignRecordsSheet.js:79]: the avatar is not being set in state right? couldn't find the records being updated here maybe I'm missing something, same for the cover

File: src/screens/ENSConfirmRegisterSheet.js [hunk 5]
```diff
@@ -137,14 +137,22 @@ export default function ENSConfirmRegisterSheet() {
...
                     height={{ custom: avatarSize }}
                     shadow="12px heavy accent"
                     width={{ custom: avatarSize }}
-                  />
+                  >
+                    <Box
+                      as={ImgixImage}
+                      borderRadius={avatarSize / 2}
+                      height={{ custom: avatarSize }}
+                      source={{ uri: records.avatar }}
```
Comment (estebanmino) [inline src/screens/ENSConfirmRegisterSheet.js:152]: ah I do see it here, getting the avatar from the records

File: src/handlers/pinata.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      {
+        filename,
+        filepath: path,
+        filetype: mime,
+        name: 'file',
+      },
+    ],
+    headers: {
+      pinata_api_key: PINATA_API_KEY,
+      pinata_secret_api_key: PINATA_API_SECRET,
```
Comment (estebanmino) [inline src/handlers/pinata.ts:43]: should we create a rainbow api key / secret?

Comment (jxom) [reply src/handlers/pinata.ts:43]: wym? this api key & secret is coming from rainbow's pinata

File: src/components/ens-registration/Avatar/Avatar.tsx (hunk 3 above)
Comment (jxom) [reply src/components/ens-registration/Avatar/Avatar.tsx:63]: ya!!

Comment (jxom) [reply src/components/ens-registration/Avatar/Avatar.tsx:63]: actually, thinking about this, does it make sense to move it there? this SVG is not an icon

File: src/handlers/pinata.ts (hunk 6 above)
Comment (estebanmino) [reply src/handlers/pinata.ts:43]: hm that I don't see it on our dev keys, we need to add it there too probably from a rainbow team account

Comment (estebanmino) [reply src/handlers/pinata.ts:43]: if you did already we need to push it to our dev keys

File: src/components/ens-registration/Avatar/Avatar.tsx (hunk 3 above)
Comment (estebanmino) [reply src/components/ens-registration/Avatar/Avatar.tsx:63]: maybe under src/components/svg/ ?

File: src/hooks/useSelectImageMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  const handleSelectImage = useCallback(async () => {
+    const image = await ImagePicker.openPicker({
+      ...imagePickerOptions,
+      includeBase64: true,
+      mediaType: 'photo',
+    });
+    onChangeImage?.({ image });
+
+    if (uploadToIPFS) {
```
Comment (estebanmino) [inline src/hooks/useSelectImageMenu.tsx:72]: i was trying this and I think we don't want to upload the images every time the user selects one, if the image is changed many times or if the user sets an image but doesn't register the ENS we'll upload bunch of images that won't be used

we probably want to upload just before sending the multicall tx to send the tx with the image link, WDYT?

Comment (jxom) [reply src/hooks/useSelectImageMenu.tsx:72]: yeah thats true. maybe we can add this as a note for now, and then implement once we have the designs for the multicall tx finalised?

File: src/components/ens-registration/Avatar/Avatar.tsx (hunk 3 above)
Comment (jxom) [reply src/components/ens-registration/Avatar/Avatar.tsx:63]: sounds good!

**Review by estebanmino — APPROVED** [review]

**Pull request merged**